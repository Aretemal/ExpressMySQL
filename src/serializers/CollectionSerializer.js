import fullUrlCreator from '../utils/fullUrlCreator.js';
import Serializer from './Serializer.js';

class CollectionSerializer extends Serializer {
  constructor(resource, serializerType, request = null, metaData = null) {
    super(resource, request);
    this.serializerType = serializerType;
    this.metaData = metaData;
  }

  meta() {
    return this.metaData;
  }

  serialize() {
    const data = {
      data: this.collect(),
      links: this.links(),
    };
    if (this.metaData) {
      data.meta = this.meta();
    }
    return data;
  }

  collect() {
    return this.resource.map((item) => {
      const serializer = new this.serializerType(item);
      return serializer.serialize().data;
    });
  }

  links() {
    return { self: `${process.env.API_URL}/${this.request.originalUrl}` };
  }
}
export default CollectionSerializer;
