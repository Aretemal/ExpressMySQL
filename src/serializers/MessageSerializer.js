import { getClass } from '../utils/getClass.js';
import Serializer from './Serializer.js';

class MessageSerializer extends Serializer {
  type() {
    return 'Message';
  }

  serialize() {
    const data = {
      data: {
        id: this.id(),
        type: this.type(),
        attributes: this.attributes(),
        links: this.links(),
      },
    };
    if (this.include) {
      data.data.relationships = this.relationships();
    }
    return data;
  }

  relationships() {
    const data = {};
    this.include.forEach((relationName) => {
      const RelationClass = this.getter(relationName);
      const serializer = new RelationClass(this.resource[relationName], {
        include: this.include,
      });
      data[relationName] = serializer.createRelationUnit();
    });
    return data;
  }

  attributes() {
    return {
      senderId: this.resource.senderId,
      dialogId: this.resource.dialogId,
      message: this.resource.message,
    };
  }

  links() {
    return { self: `${process.env.API_URL}/dialog/message/${this.resource.id}` };
  }
}
export default MessageSerializer;
