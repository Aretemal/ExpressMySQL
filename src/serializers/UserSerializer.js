import Serializer from './Serializer.js';

class UserSerializer extends Serializer {
  type() {
    return 'User';
  }

  id() {
    return this.resource.userId;
  }

  collectionSerializer() {
    return {
      type: this.type(),
      id: this.id(),
      attributes: this.attributes(),
    };
  }

  attributes() {
    return {
      fullName: this.resource.fullName,
      lastName: this.resource.lastName,
      login: this.resource.login,
      status: this.resource.status,
      email: this.resource.email,
      ava: this.resource.ava,
    };
  }
}
export default UserSerializer;
