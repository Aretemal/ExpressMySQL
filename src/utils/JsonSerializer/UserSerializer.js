import Serializer from './Serializer.js';

class UserSerializer extends Serializer {
  getOne(data, links) {
    return this.serializer(data, data.id, 'ObjectUser', links);
  }

  getAllUsers(data, links) {
    return this.serializer(data, data.userAuth.id, 'Array Users', links);
  }

  getStatus(data, links) {
    return this.serializer(data, data.id, 'Status', links);
  }
}
export default new UserSerializer();
