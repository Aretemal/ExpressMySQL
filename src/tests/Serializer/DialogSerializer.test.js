import { matchersWithOptions } from 'jest-json-schema';
import CollectionSerializer from '../../serializers/CollectionSerializer.js';
import MessageSerializer from '../../serializers/MessageSerializer.js';
import UserSerializer from '../../serializers/UserSerializer.js';
import schema from './schema.json';

expect.extend(matchersWithOptions({
  verbose: true,
}));

describe('MessageSerializer :', () => {
  describe('sendMessage :', () => {
    test('should return created message', () => {
      const message = {
        senderId: 1,
        recipientId: 2,
        message: 'Hello',
      };

      const serializer = new MessageSerializer(message);
      const data = serializer.serialize();

      expect(data).toMatchSnapshot();
      expect(data).toMatchSchema(schema);
    });
  });

  describe('getAllMessage :', () => {
    test('should return array with messages', () => {
      const messages = [{
        senderId: 1,
        message: 'Hello',
      },
      {
        senderId: 2,
        message: 'Hel1lo',
      }];
      const users = [
        {
          id: 1,
          login: 'Artem',
          firstName: 'aaaa',
          lastName: 'bbbb',
          email: '111@11.1',
          status: '1',
          ava: null,
        }, {
          id: 2,
          login: 'Ar3tem',
          firstName: 'a4aaa',
          lastName: 'bb5bb',
          email: '111@161.1',
          status: '16',
          ava: null,
        },
      ];

      const serializer = new CollectionSerializer(
        messages,
        MessageSerializer,
        { originalUrl: 'example' },
        null,
        users,
      );
      const data = serializer.serialize();

      expect(data).toMatchSnapshot();
      // expect(data).toMatchSchema(schema);
    });
  });

  describe('getAllInterlocutors :', () => {
    test('should return array of users', () => {
      const users = [{
        id: 3,
        login: 'dsdsd',
        firstName: 'assaaa',
        lastName: 'bbsbb',
        email: '111s@11.1',
        status: '1s',
        ava: null,
      },
      {
        id: 1,
        login: 'Artem',
        firstName: 'aaaa',
        lastName: 'bbbb',
        email: '111@11.1',
        status: '1',
        ava: null,
      }];

      const serializer = new CollectionSerializer(users, UserSerializer, { originalUrl: 'example' });
      const data = serializer.serialize();

      expect(data).toMatchSnapshot();
      expect(data).toMatchSchema(schema);
    });
  });
});
