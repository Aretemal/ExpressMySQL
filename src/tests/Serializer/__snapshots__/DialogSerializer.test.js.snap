// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DialogSerializer : getAllInterlocutors : should return array of users 1`] = `
{
  "data": [
    {
      "attributes": {
        "ava": null,
        "email": "111s@11.1",
        "firstName": "assaaa",
        "lastName": "bbsbb",
        "login": "dsdsd",
        "status": "1s",
      },
      "id": "3",
      "links": {
        "self": "http://localhost:5000/api/user/3",
      },
      "type": "User",
    },
    {
      "attributes": {
        "ava": null,
        "email": "111@11.1",
        "firstName": "aaaa",
        "lastName": "bbbb",
        "login": "Artem",
        "status": "1",
      },
      "id": "1",
      "links": {
        "self": "http://localhost:5000/api/user/1",
      },
      "type": "User",
    },
  ],
  "links": {
    "self": "http://localhost:5000/api/example",
  },
}
`;
